jax/experimental/sparse/linalg.py

```diff
@@ -0,0 +1,388 @@
+      'P.X': PX}
+
+def _eigh_ascending(A):
+  w, V = jnp.linalg.eigh(A)
+  return w[::-1], V[:, ::-1]
+
+
+def _svqb(X):
+  # https://sdm.lbl.gov/~kewu/ps/45577.html
+  # May choose to truncate the basis rather than return
+  # a degenerate, non-orthogonal one.
+
+  norms = jnp.linalg.norm(X, ord=2, axis=0, keepdims=True)
+  X /= jnp.where(norms == 0, 1.0, norms)
+
+  inner = _mm(X.T, X)
+
+  w, V = _eigh_ascending(inner)
+
+  tau = jnp.finfo(X.dtype).eps * w[0]
+  padded = jnp.maximum(w, tau)
+  sqrted = jnp.where(tau > 0, padded, 1.0) ** (-0.5)
+  scaledV = V * sqrted[jnp.newaxis, :]
+  orthoX = _mm(X, scaledV)
+
+  keep = ((w > tau) * (jnp.diag(inner) > 0.0))[jnp.newaxis, :]
+  orthoX *= keep
+  norms = jnp.linalg.norm(orthoX, ord=2, axis=0, keepdims=True)
+  keep *= norms > 0.0
+  orthoX /= jnp.where(keep, norms, 1.0)
+  return orthoX
+
+
+def _project_out(basis, U):
+  # "twice is enough" from shoyer's reference:
+  # http://slepc.upv.es/documentation/reports/str1.pdf
+
+  for _ in range(2):
+    U -= _mm(basis, _mm(basis.T, U))
+    U = _orthonormalize(U)
+
+  # It's crucial to end on a subtraction of the original basis.
+  # This seems to be a detail not present in [2], possibly because of
+  # of reliance on soft locking.
+  #
+  # Near convergence, if the residuals R are 0 and our last
+  # operation when projecting (X, P) out from R is the orthonormalization
+  # done above, then due to catastrophic cancellation we may re-introduce
+  # (X, P) subspace components into U, which can ruin the Rayleigh-Ritz
+  # conditioning.
+  #
+  # We nuke any columns that are even remotely suspicious, so the invariant
+  # that [basis, U] is zero-or-orthogonal is ensured.
+  U -= _mm(basis, _mm(basis.T, U))
+  normU = jnp.linalg.norm(U, ord=2, axis=0, keepdims=True)
+  U *= normU >= 0.99
+
+  return U
+
+def _orthonormalize(basis):
+  # Twice is enough, again.
+  for _ in range(2):
+    basis = _svqb(basis)
+  return basis
+
+
+def _rayleigh_ritz_orth(A, S):
+  # Classical Rayleigh-Ritz returns w, V satisfying
+  # (1) S.T A S @ V ~= w * V
+  # such that (2) V is (S.T S)-orthonormal.
+  # https://www.netlib.org/lapack/lug/node54.html
+  #
+  # Usually it requires solving the complicated standard eigensystem
+  # U^-T S^T A S U^-1 @ Q = w * Q and then backsolving V = U^-1 Q,
+  # but if S is standard orthonormal then we just need to find
+  # eigenvalues of S.T A S.
+
+  SAS = _mm(S.T, A(S))
+  SAS = (SAS + SAS.T) / 2
+
+  # Solve the projected subsytem.
+  # If we could tell to eigh to stop after first k, we would.
+  return _eigh_ascending(SAS)
+
+
+def _extend_basis(X, m):
+  # Use a block householder reflector to generate orthogonal extension
+  # to X. There's nothing too special about this, and we could choose
+  # any random extension to X's basis, but this is a deterministic choice.
+  #
+  # https://epubs.siam.org/doi/abs/10.1137/0725014
+  # https://www.jstage.jst.go.jp/article/ipsjdc/2/0/2_0_298/_article
+  n, k = X.shape
```

Comment:
Please write out the math here instead of just citing the papers. The code is not particularly readable with all the concatenations etc.

Reply:
done